blog/2020/09/rr-memory-magic.md

```diff
@@ -0,0 +1,503 @@
+ (task 17107 (rec:7760) at time 51911)                                                                                 
+ -> Assertion `false' failed to hold. Replaying `SIGNAL: SIGSEGV(det)': expecting tracee signal or trap, but instead at
+ `write' (ticks: 144040993972)                                                                                         
+Tail of trace dump:                                                                                                    
+[snip]                                                                                                                 
+  { syscall:'clock_gettime', ret:0x0, size:0x20 }                                                                      
+}                                                                                                                      
+{                                                                                                                      
+  real_time:38257.585221 global_time:51911, event:`SIGNAL: SIGSEGV(det)' tid:7760, ticks:143998759917                  
+rax:0x436487c0 rbx:0x7f0490b74330 rcx:0xb479 rdx:0x435f2dc8 rsi:0x435f2dc0 rdi:0x435f2dc0 rbp:0x7fff6ca71250 rsp:0x7fff
+6ca71230 r8:0x1 r9:0x7f0490b74330 r10:0x1007f0490b784d0 r11:0x4364d190 r12:0x0 r13:0x7f0490b88c70 r14:0x7f047daf36d0 r1
+5:0xb479 rip:0x7f049e2ba8e3 eflags:0x10287 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0xffffffffffffffff fs_b
+ase:0x7f0499f49240 gs_base:0x0                                                                                         
+}                                                                                                                      
+{                                                                                                                      
+  real_time:38257.585223 global_time:51912, event:`SYSCALLBUF_RESET' tid:7760, ticks:143998759917                      
+}           
+```
+
+This is the point where Tim asked me to take over, since that really isn't supposed to happen.
+Now, the error message here may look a bit scary, but let's parse it out part by part. The actual error message
+is:
+
+```raw
+Replaying `SIGNAL: SIGSEGV(det)': expecting tracee signal or trap, but instead at
+ `write' (ticks: 144040993972)
+```
+
+What does this mean? First a quick refresher about how rr works: Basically, it records any modifications made to the process memory by the kernel (or any other source of non-determinism). Then, between such modification it relies on the determinism of the processor to produce a bitwise-identical output state, given an input memory state.
+
+Now, what happened here is that rr expected to get a deterministic (`(det)`) segfault (i.e. one caused by the execution of an instruction, as opposed to being asynchronously sent by some other process e.g. using `kill -SIGSEGV`). However, instead of seeing this segfault, we ended up somewhere else. In this case (`but instead at
```

Comment:
```suggestion
Now, what happened here is that `rr` expected to get a deterministic (`(det)`) segfault (i.e. one caused by the execution of an instruction, as opposed to being asynchronously sent by some other process e.g. using `kill -SIGSEGV`). However, instead of seeing this segfault, we ended up somewhere else. In this case (`but instead at
```